src/pytorch_lightning/utilities/auto_restart.py

```diff
@@ -268,17 +296,18 @@ def __init__(self, dataset: IterableDataset) -> None:
         super().__init__()
         self.dataset = deepcopy(dataset)
         self.samplers: Optional[Dict[str, FastForwardSampler]] = None
-        self._state_dict: Optional[Dict[int, Any]] = None
+        self._state_dict: Optional[Dict[str, Any]] = None
         self._has_wrapped: bool = False
 
     @property
     def sampler(self) -> Sampler:
         return self.dataset.sampler
 
     def state_dict(self) -> Dict[str, Any]:
+        assert self.samplers is not None
         return {k: v.state_dict() for k, v in self.samplers.items()}
 
-    def load_state_dict(self, state_dict: Dict[int, Any]) -> None:
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
```

Comment:
If this is getting changed, then why do we introduce the IntStateful type? I'm a little confused about this.

Reply:
The `_IntStateful` was added because some input of `load_state_dict` is `Dict[int, Any]`. In particular,
https://github.com/Lightning-AI/lightning/blob/f3ef4f8b4d5231442b0305997da718a2684040e0/src/pytorch_lightning/utilities/auto_restart.py#L604-L605

The output of `_rotate_worker_indices` is expected to be `Dict[int, Any]`.
https://github.com/Lightning-AI/lightning/blob/f3ef4f8b4d5231442b0305997da718a2684040e0/src/pytorch_lightning/utilities/auto_restart.py#L626